### PR TITLE
Add method completion callback for basic.consume

### LINKED ISSUE #953: Channel.basic_consume doesn't support consume-ok/completion callback
[Channel.basic_consume](https://github.com/pika/pika/blob/3d3b95d31b67dfeaf5ef43650c162e25169336e6/pika/channel.py#L263) doesn't provide an argument for specifying the completion callback to be called upon [consume-ok](https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume-ok). (the callback arg that it presently supports is for consuming incoming messages).

A client's logic might need to know that the `basic-consume` request succeeded before initiating other operations (e.g., when requesting exclusive access). Lack of the completion callback prevents the application from doing so easily.

Implementation notes:
1. Rename `Channel.basic_consume`'s consumer callback (presently named `callback`) to `consumer_callback`, `message_callback`, `message_rx_callback`, or similar, and name the completion callback `callback` to follow the recently-established convention for naming of completion callbacks. Also rename the consumer callback arg in `BlockingChannel.basic_consume` for consistency with `Channel.basic_consume`.
2. If completion callback isn't specified, auto-generate and return the consumer tag  following the legacy behavior. If completion callback is specified, return None from `Channel.basic_consume` and let the server generate the consumer tag and return it via [consume-ok](https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume-ok). Update the docstring in `Channel.basic_consume` to reflect this.
3. Make sure pika's [Channel._generate_consumer_tag](https://github.com/pika/pika/blob/3d3b95d31b67dfeaf5ef43650c162e25169336e6/pika/channel.py#L323) doesn't generate consumer tags that can potentially collide with server-generated ones by prefixing "pika" into the consumer tag name that `Channel._generate_consumer_tag` generates.

### PR DESCRIPTION
**Pull request by lukebakken**
Fixes #953

### CODE REVIEW

File: tests/unit/channel_tests.py [hunk 1]
```diff
@@ -243,73 +243,94 @@ def test_basic_cancel_unknown_consumer_tag(self, rpc):
...
     @mock.patch('pika.channel.Channel._validate_channel')
-    def test_basic_consume_calls_validate(self, validate):
+    @mock.patch('pika.channel.Channel._require_callback')
+    def test_basic_consume_calls_validate(self, require, validate):
         self.obj._set_state(self.obj.OPEN)
         mock_callback = mock.Mock()
-        self.obj.basic_consume('test-queue', mock_callback)
+        mock_on_msg_callback = mock.Mock()
+        self.obj.basic_consume('test-queue', mock_on_msg_callback,
+                callback=mock_callback)
```
Comment (vitaly-krugl) [inline tests/unit/channel_tests.py]: This change and several others trigger "Wrong continued indentation" pylint warnings

File: pika/adapters/blocking_connection.py [hunk 2]
```diff
@@ -1541,30 +1541,30 @@ def basic_consume(self,
             consumer_tag is already present.
 
         """
-        if not callable(callback):
-            raise ValueError('callback callback must be callable; got %r'
+        if not callable(on_message_callback):
+            raise ValueError('callback on_message_callback must be callable; got %r'
                              % callback)
```
Comment (vitaly-krugl) [inline pika/adapters/blocking_connection.py]: Per pylint: "Undefined variable 'callback'". Sounds like we are missing a unit test for `not callable(on_message_callback)`.

**Comment by vitaly-krugl** [comment]
@lukebakken, I wanted to push a small PR to your fork/branch, but couldn't find the source fork and branch of your PR, and couldn't find how to do it by googling. Do you know how to discover it in github?

In the interest of time, I added a few lines to tests/unit/channel_tests.py disabling several pylint warnings that were responsible for a bunch of pylint output that we don't care about in test modules, which made it difficult to see issues that we actually care about. Here is the change:
```
$ git diff @~..@
diff --git a/tests/unit/channel_tests.py b/tests/unit/channel_tests.py
index 2c6ff5c..85446a5 100644
--- a/tests/unit/channel_tests.py
+++ b/tests/unit/channel_tests.py
@@ -12,6 +12,10 @@ import mock
 
 from pika import channel, connection, exceptions, frame, spec
 
+# Disable protected-access, missing-docstring, and invalid-name,
+# too-many-public-methods, too-many-lines
+# pylint: disable=W0212,C0111,C0103,R0904,C0302
+
```

**Comment by lukebakken** [comment]
> I wanted to push a small PR to your fork/branch, but couldn't find the source fork and branch of your PR, and couldn't find how to do it by googling. Do you know how to discover it in github?

@vitaly-krugl - I don't use a fork, and just push branches to the `github.com/pika/pika` repository. Always feel free to make changes to my branches if you'd like. This branch is `pika-953`

File: tests/acceptance/blocking_adapter_test.py [hunk 3]
```diff
@@ -1564,12 +1563,8 @@ def test(self):  # pylint: disable=R0914,R0915
 
         # Create a consumer
         rx_messages = []
-        consumer_tag = ch.basic_consume(
-            q_name,
-            lambda *args: rx_messages.append(args),
-            no_ack=False,
-            exclusive=False,
-            arguments=None)
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply tests/acceptance/blocking_adapter_test.py]: Sure, I can revert this.

**Comment by vitaly-krugl** [comment]
@lukebakken, you don't like Git Workflow?

**Comment by lukebakken** [comment]
I don't see a reason to use a fork if I have push access to a repository is all 😸

File: tests/unit/channel_tests.py [hunk 4]
```diff
@@ -243,73 +243,94 @@ def test_basic_cancel_unknown_consumer_tag(self, rpc):
...
         self.obj.basic_consume(
-            'test-queue', mock_callback, consumer_tag=consumer_tag)
+            'test-queue', mock_on_msg_callback,
+            consumer_tag=consumer_tag, callback=mock_callback)
         expectation = spec.Basic.Consume(
             queue='test-queue',
             consumer_tag=consumer_tag,
             no_ack=False,
             exclusive=False)
-        rpc.assert_called_once_with(expectation, self.obj._on_eventok,
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply tests/unit/channel_tests.py]: OK, getting that now

File: tests/unit/channel_tests.py (hunk 1 above)
Comment (lukebakken) [reply tests/unit/channel_tests.py]: Blame vim!!! 😉

**Comment by lukebakken** [comment]
Well, there is that. You used to not be able to add commits to branches on forks, either, but that has changed. Maybe next PR.

File: tests/unit/channel_tests.py (hunk 1 above)
Comment (vitaly-krugl) [reply tests/unit/channel_tests.py]: Bad vim, very bad vim! :)

**Comment by vitaly-krugl** [comment]
One of the nice things about forks is that other people can push PRs to your fork/branch, and you can accept/reject them, as I wanted to do above.

**Comment by vitaly-krugl** [comment]
Okay, please ping me when you're ready for the final review.

**Comment by lukebakken** [comment]
@vitaly-krugl - all set, thanks!

**Pull request merged**